test/e2e/storage/testsuites/provisioning.go

```diff
@@ -306,6 +316,95 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
 	return pv
 }
 
+// PVWriteReadCheck checks that a PV retains data.
+//
+// It starts these pods:
+// - The first pod writes 'hello word' to the /mnt/test (= the volume) on one node.
+// - The second pod runs grep 'hello world' on /mnt/test on the same node.
+// - Optional: a third pod runs grep 'hello world' on /mnt/test on a different node.
+//
+// Nodes are selected like this:
+// - the first node is based on nodeName (if not empty) or nodeSelector
+//   (otherwise), random if both are empty
+// - the second node is only used if nodeSelector is set and matches more than
+//   the first node
+// - if both nodeName and nodeSelector are unset, then the test cannot lock
+//   the pod onto the same node
+//
+// If all succeeds, Kubernetes actually allocated something that is
+// persistent across pods and (optional) across nodes.
+//
+// This is a common test that can be called from a StorageClassTest.PvCheck.
+func PVWriteReadCheck(client clientset.Interface, claim *v1.PersistentVolumeClaim, volume *v1.PersistentVolume, nodeName string, nodeSelector map[string]string) {
+	firstNodeName, nodes := pickNode(client, nodeName, nodeSelector)
+
+	// We keep the node selection logic in Kubernetes active by always using a node selector
+	// although we have already picked one node.
+	var firstNodeSelector map[string]string
+	if firstNodeName != "" {
+		firstNodeSelector = map[string]string{"kubernetes.io/hostname": firstNodeName}
+		By(fmt.Sprintf("checking the created volume is writable and has the PV's mount options on node %q", firstNodeName))
+	} else {
+		By("checking the created volume is writable and has the PV's mount options on some node")
+	}
+	command := "echo 'hello world' > /mnt/test/data"
+	// We give the first pod the secondary responsibility of checking the volume has
+	// been mounted with the PV's mount options, if the PV was provisioned with any
+	for _, option := range volume.Spec.MountOptions {
+		// Get entry, get mount options at 6th word, replace brackets with commas
+		command += fmt.Sprintf(" && ( mount | grep 'on /mnt/test' | awk '{print $6}' | sed 's/^(/,/; s/)$/,/' | grep -q ,%s, )", option)
+	}
+	command += " || (mount | grep 'on /mnt/test'; false)"
+	runInPodWithVolume(client, claim.Namespace, claim.Name, "-first", "" /* node name */, command, firstNodeSelector, false)
+
+	if firstNodeName != "" {
+		By(fmt.Sprintf("checking the created volume is readable and retains data on the same node %q", firstNodeName))
+	} else {
+		By("checking the created volume is readable and retains data on some node")
+	}
+	command = "grep 'hello world' /mnt/test/data"
+	runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", "" /* node name */, command, firstNodeSelector, false)
+
+	// Run on another node, if we have one.
+	var secondNodeName string
+	if nodes != nil {
+		for _, node := range nodes.Items {
```

Comment:
In a multizone cluster nodes are going to be in multiple zones.  Blindly picking any node could conflict with PV.nodeaffinity.

I would suggest instead setting anti affinity to the first node as a way to pick the second node.

Reply:
The intended usage for this test is so that the node selector matches all nodes that are compatible with the PV. I guess need to spell this out more explicitly.

When used like that, the problem you mentioned cannot happen, can it?

Reply:
The node selectors were originally a workaround when we didn't have a concept of topology for PVs and we needed to restrict the pods to specific zones.  Now that we have PV.NodeAffinity, we shouldn't need to restrict pods with a node selector.

Reply:
As a side note, topology in CSI is still alpha and disabled by default.  For now, we have disabled multi-zone CSI GCE PD tests until topology goes beta.  So I think we should still design the test case around leveraging PV topology.

Reply:
The tests no longer pick a node themselves. Instead I let Kubernetes schedule a pod, then determine where it landed to schedule the second pod (by name for the "same node" test, by anti-affinity for the "other node"  test).

The provisioning test was only using the node name and ignored the node selector that was introduced for the topology tests. I've kept it that way.